### PR TITLE
Public page template response

### LINKED ISSUE #6553: Add public page template for apps
Many apps now offer public pages that are shown to users that are not logged in to the instance.
There should be a common template for those pages, that apps can use.

Right now every app includes its own template:
- gallery: https://github.com/nextcloud/gallery/blob/master/templates/public.php
- calendar: https://github.com/nextcloud/calendar/blob/master/templates/public.php
- files_sharing: https://github.com/nextcloud/server/blob/master/apps/files_sharing/templates/public.php
- spreed: https://github.com/nextcloud/spreed/blob/master/templates/index-public.php
- polls needs that as well: https://github.com/nextcloud/polls/issues/96#issuecomment-330154123

@nextcloud/designers

### PR DESCRIPTION
**Pull request by juliusknorr**
This PR is a proof of concept to fix #6553 

The header bar should be the same across all public pages of apps, therefore I tried to add a simple abstraction layer to that. We should stop app developers messing to much with custom HTML and therefore it is now possible to set the following parameters for the header:
- Title (Filename in files_sharing public page)
- Details ("shared by user" in files_sharing public page)
- Menu actions (will show the single action with icon or the first with a 3-dots menu for the other actions)

Example usage is shown for files_sharing https://github.com/nextcloud/server/compare/public-template#diff-d6e103c0f5540d507d9c9b6c7ade2151R432

Actions are a simple class containing the icon, label, link and a possible details label, but apps can add their own types by implementing an IMenuAction.

I'm still not sure if that abstraction is to much, but I like having a simple interface to setup those basic things from an app developers perspective. ~Maybe we need to move out the HTML from the IMenuAction implementations to some files in the template directories~. I decided for now to keep the templating code inside the classes, since otherwise we would have to load a template file for each menu entry. We possibly can improve on that later. Ideally there should be some template engine providing proper caching for templates.

Some feedback would be nice @rullzer @MorrisJobke @ChristophWurst @nickvergessen 
maybe also from @nextcloud/designers if there are any other default components we should have for public pages.

With one action added:
bildschirmfoto vom 2018-01-25 20-40-09

With multiple actions added:
bildschirmfoto vom 2018-01-25 20-39-27

### CODE REVIEW

**Comment by MorrisJobke** [comment]
In general I like this idea 👍 Haven't had a look at the code.

**Comment by rullzer** [comment]
Awesome idea @juliushaertl!
Reminds me that I should continue https://github.com/nextcloud/server/pull/6760 as they complement each other nicely :)

**Comment by oparoz** [comment]
Make sure scrolling still works on iOS Safari when having long pages. Last time the structure was changed, performance was unacceptable.

**Comment by juliushaertl** [comment]
@oparoz Do you have any reference to the last change there, so I can have a look what should be done? Otherwise testing by any iOS user is appreciated. :wink: The only change from the HTML layout perspective is affecting the menu actions in the top right. Everything else like the rendering of the files_sharing pages should stay untouched.

**Comment by juliushaertl** [comment]
Hm, acceptance tests are still failing. I'll have a look.

**Comment by pixelipo** [comment]
Is this ready for review or still in Development, @juliushaertl ?

**Comment by juliushaertl** [comment]
@pixelipo Ah forgot to mention, ready for review. :wink:

File: lib/public/AppFramework/Http/Template/SimpleMenuAction.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		$this->link = $link;
+		$this->priority = $priority;
+		$this->detail = $detail;
+	}
+
+	/**
+	 * @param string $id
+	 * @since 14.0.0
+	 */
+	public function setId(string $id) {
```
Comment (MorrisJobke) [inline lib/public/AppFramework/Http/Template/SimpleMenuAction.php]: Are all those setters really needed? Isn't the stuff in the constructor fine?

**Review by MorrisJobke — APPROVED** [review]
Fine by me - only have one little comment.

File: lib/public/AppFramework/Http/Template/SimpleMenuAction.php (hunk 1 above)
Comment (juliushaertl) [reply lib/public/AppFramework/Http/Template/SimpleMenuAction.php]: Hm right, there's no real need for those. Removed them.

**Review by rullzer — APPROVED** [review]

**Comment by skjnldsv** [comment]
Awesome!! :100: 
To the documentation now @juliushaertl :rocket: :laughing:

**Comment by desaintmartin** [comment]
Hello,
This commit seems to break the public page of a shared folder in the Gallery app.

Here is my log using the latests commits from server.git and gallery.git :

```
    Message: Call to a member function getHeaderTitle() on null
    File: /srv/www/nextcloud-dev-test.watercooled.org/core/templates/layout.public.php
    Line: 36


Trace

#0 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/Template/Base.php(179): include()
#1 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/Template/Base.php(151): OC\Template\Base->load('/srv/www/nextcl...', Array)
#2 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/legacy/template.php(204): OC\Template\Base->fetchPage(Array)
#3 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/legacy/template.php(235): OC_Template->fetchPage(Array)
#4 /srv/www/nextcloud-dev-test.watercooled.org/lib/public/AppFramework/Http/TemplateResponse.php(157): OC_Template->fetchPage(Array)
#5 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/AppFramework/Http/Dispatcher.php(119): OCP\AppFramework\Http\TemplateResponse->render()
#6 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/AppFramework/App.php(117): OC\AppFramework\Http\Dispatcher->dispatch(Object(OCA\Gallery\Controller\PageController), 'publicIndex')
#7 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/AppFramework/Routing/RouteActionHandler.php(47): OC\AppFramework\App::main('PageController', 'publicIndex', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)
#8 [internal function]: OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)
#9 /srv/www/nextcloud-dev-test.watercooled.org/lib/private/Route/Router.php(297): call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)
#10 /srv/www/nextcloud-dev-test.watercooled.org/lib/base.php(981): OC\Route\Router->match('/apps/gallery/s...')
#11 /srv/www/nextcloud-dev-test.watercooled.org/index.php(37): OC::handleRequest()
#12 {main}


```

**Comment by rullzer** [comment]
@juliushaertl and to the gallery it seems ;)

**Comment by juliushaertl** [comment]
I'll prepare some follow up pull requests for other apps that use public pages.

**Comment by juliushaertl** [comment]
From a first look int seems that gallery is using 'public' as layout, which was not defined before this PR. Chaning the last argument from this line https://github.com/nextcloud/gallery/blob/2005c66ab59a9dcd66d6794834434750eabea792/lib/Controller/PageController.php#L235 to 'base' should work around the issue until I prepared a proper PR to move to the new public layout.

**Pull request merged**